lib/OpenQA/Utils.pm

```diff
@@ -254,7 +255,27 @@ sub is_in_tests {
 sub needledir { productdir(@_) . '/needles' }
 
 sub locate_needle {
-    my ($relative_needle_path, $needles_dir) = @_;
+    my ($relative_needle_path, $needles_dir, $needles_ref) = @_;
+
+    if ($needles_ref) {
+        my $needles_dir_basename = basename(dirname($needles_dir));
+        my $temp_needles_dir = "/tmp/needle_dirs/$needles_dir_basename/$needles_ref/needles";
+        if (File::Spec->splitdir($relative_needle_path) > 1) {
+            make_path($temp_needles_dir . '/' . dirname($relative_needle_path));
+        }
+        my $temp_json_path = $temp_needles_dir . '/' . $relative_needle_path;
+        open my $temp_json_fh, '>', $temp_json_path;
+        my $jsonfile = qx{git -C $needles_dir show $needles_ref:./$relative_needle_path};
+        print $temp_json_fh $jsonfile;
+        close $temp_json_fh;
+        my $png_name = dirname($relative_needle_path) . '/' . basename($relative_needle_path, '.json') . '.png';
+        my $temp_png_path = $temp_needles_dir . '/' . $png_name;
+        open my $temp_png_fh, '>', $temp_png_path;
+        my $pngfile = qx{git -C $needles_dir show $needles_ref:./$png_name};
```

Comment:
We should handle the case when this call fails (Git returns a non-zero exit status or cannot be invoked at all).

Note that error handling for `qx{…}` calls is a bit involved: https://perldoc.perl.org/perlop#Quote-Like-Operators
So you might want to introduce a helper for this (maybe we already have one).